core/primitives/src/time.rs

```diff
@@ -0,0 +1,156 @@
+use std::default::Default;
+use std::sync::{Arc, Mutex};
+
+use chrono;
+use once_cell::sync::Lazy;
+
+pub use chrono::Utc;
+pub use std::time::{Duration, Instant};
+
+use chrono::DateTime;
+use std::collections::{HashMap, VecDeque};
+use std::ops::Deref;
+use std::thread::ThreadId;
+
+pub struct MockClockPerThread {
+    utc: VecDeque<DateTime<Utc>>,
+    durations: VecDeque<Duration>,
+    utc_call_count: u64,
+    instant_call_count: u64,
+    instant: Instant,
+    is_mock: bool,
+}
+
+pub struct Clock {
+    clocks_per_thread: HashMap<ThreadId, MockClockPerThread>,
+}
+
+impl MockClockPerThread {
+    pub fn reset(&mut self) {
+        self.utc.clear();
+        self.durations.clear();
+        self.utc_call_count = 0;
+        self.instant_call_count = 0;
+        self.instant = Instant::now();
+        self.is_mock = false;
+    }
+}
+
+impl Default for MockClockPerThread {
+    fn default() -> Self {
+        Self {
+            utc: VecDeque::with_capacity(16),
+            durations: VecDeque::with_capacity(16),
+            utc_call_count: 0,
+            instant_call_count: 0,
+            instant: Instant::now(),
+            is_mock: false,
+        }
+    }
+}
+
+impl Default for Clock {
+    fn default() -> Self {
+        Self { clocks_per_thread: HashMap::default() }
+    }
+}
+
+static SINGLETON: Lazy<Arc<Mutex<Clock>>> = Lazy::new(|| Arc::from(Mutex::new(Clock::new())));
```

Comment:
Rather than maintaining per-thread map, Arc and lazy ourselves, we can leverage std::therad_local:

```rust
struct MockClockPerThread {
    utc: VecDeque<DateTime<Utc>>,
    durations: VecDeque<Duration>,
    utc_call_count: u64,
    instant_call_count: u64,
    instant: Instant,
    is_mock: bool,
}

impl MockClockPerThread {
    fn with<F, T>(f: F) -> T
    where
        F: FnOnce(&mut MockClockPerThread) -> T,
    {
        thread_local! {
            static INSTANCE: RefCell<MockClockPerThread> = RefCell::default()
        }
        INSTANCE.with(|it| f(&mut *it.borrow_mut()))
    }

    fn pop_utc(&mut self) -> Option<DateTime<chrono::Utc>> {
        self.utc_call_count += 1;
        self.utc.pop_front()
    }
}

pub struct Clock {}

impl Clock {
    pub fn add_utc(mock_date: DateTime<chrono::Utc>) {
        MockClockPerThread::with(|clock| {
            clock.utc.push_back(mock_date);
            clock.is_mock = true;
        });
    }
    pub fn utc() -> DateTime<chrono::Utc> {
        MockClockPerThread::with(|clock| {
            let x = clock.pop_utc();
            match x {
                Some(t) => t,
                None => {
                    if clock.is_mock {
                        panic!("Mock clock run out of samples");
                    }
                    chrono::Utc::now()
                }
            }
        })
    }
}
```

Reply:
To clarify -- the current design also works I think, no need to rewrite if you don't like the new version better.

Reply:
:thinking: that being said, I think there's a potential problem with the hash map -- we never remove entries form it. If we do maintain hashmap in the prod code, and we do create new threads from time to time, than this would leak memory. If that is indeed the problem (which I dont' know), we should switch to `thread_local`, as that includes the necessary code to release resources when a thread exists.

Reply:
Done :)